### PR TITLE
Cross domain UCR: report builder changes

### PR DESCRIPTION
**Pull request by stephherbers**
## Summary

Ticket: https://dimagi-dev.atlassian.net/browse/USH-1140

## Product Description

modify ConfigureReport and adds RegistryCaseDataSourceHelper

## Safety Assurance

- [X] Risk label is set correctly
- [X] All migrations are backwards compatible and won't block deploy
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [X] If QA is part of the safety story, the "Awaiting QA" label is used
- [X] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage

added tests for new RegistryCaseDataSourceHelper in DataSourceReferenceTest.test_reference_for_registry

### QA Plan

I will be requesting QA along with this [PR](https://github.com/dimagi/commcare-hq/pull/30362)
### Rollback instructions

- [X] This PR can be reverted after deploy with no further considerations

### CODE REVIEW

File: corehq/apps/userreports/reports/builder/forms.py
Comment (snopoke) [inline corehq/apps/userreports/reports/builder/forms.py:858]: not 100% sure where to put this but somewhere we need to check that this domain / user has access to the DR (https://github.com/dimagi/commcare-hq/blob/3c8c6e7ce8622a1f5e7d1c57f8c0f14f210c0b73/corehq/apps/registry/helper.py#L55) and that the case type is in the DR schema (https://github.com/dimagi/commcare-hq/blob/3c8c6e7ce8622a1f5e7d1c57f8c0f14f210c0b73/corehq/apps/registry/helper.py#L51)

File: corehq/apps/userreports/reports/builder/forms.py
Comment (snopoke) [inline corehq/apps/userreports/reports/builder/forms.py:875]: probably also validate that `source_type == DATA_SOURCE_TYPE_CASE`

File: corehq/apps/userreports/reports/builder/forms.py
Comment (snopoke) [inline corehq/apps/userreports/reports/builder/forms.py:864]: we should add the registry slug to this meta class so that we can use it when editing the report (see `ConfigureNewReportBase._bootstrap`)

File: corehq/apps/userreports/reports/builder/forms.py
Comment (snopoke) [inline corehq/apps/userreports/reports/builder/forms.py:1010]: This is probably the best place to validate access to the data registry (or somewhere in this class)

File: corehq/apps/userreports/tests/test_report_builder.py
Comment (snopoke) [inline corehq/apps/userreports/tests/test_report_builder.py:193]: you shouldn't need to delete SQL models in tests since they get removed when the test transaction is rolled back. However if you did want to the proper way is to use `self.addCleanup(registry.delete)` immediately after you've created it. This will get called during test tear down.

File: corehq/apps/userreports/views.py
Comment (snopoke) [inline corehq/apps/userreports/views.py:487]:
```suggestion
                self.source_type = DATA_SOURCE_TYPE_CASE
```

File: corehq/apps/userreports/views.py
Comment (snopoke) [inline corehq/apps/userreports/views.py:489]: :+1: we shouldn't need an app here

**Review by snopoke — COMMENTED** [review]
few comments but looking good. There are a number of TODOs, are you planning to address those in this PR or as a followup?

File: corehq/apps/userreports/tests/test_report_builder.py
Comment (stephherbers) [reply corehq/apps/userreports/tests/test_report_builder.py]: sweet thank you

File: corehq/apps/userreports/reports/builder/forms.py
Comment (stephherbers) [reply corehq/apps/userreports/reports/builder/forms.py:1001]: From some other comments above added the validations here: https://github.com/dimagi/commcare-hq/pull/30285/commits/9db442acc3d5aa2c3eb9c78676736312ee628d68
It feels a bit janky so very open to other suggestions

File: corehq/apps/userreports/reports/builder/forms.py
Comment (snopoke) [inline corehq/apps/userreports/reports/builder/forms.py:849]: it's easier if you let the helper load the registry for you as well since it will also do the access check then
```suggestion
        helper = DataRegistryHelper(self.domain, registry_slug=self.registry_slug)
        helper.pre_access_check(source_type)
        owning_domain = helper.registry.domain
```

File: corehq/apps/userreports/reports/builder/forms.py
Comment (snopoke) [inline corehq/apps/userreports/reports/builder/forms.py:1039]:
```suggestion
            elif not self.registry_slug:
```

File: corehq/apps/userreports/views.py
Comment (snopoke) [inline corehq/apps/userreports/views.py:492]:
```suggestion
                helper = DataRegistryHelper(self.domain, registry_slug=self.registry_slug)
                helper.check_access(request.couch_user)
```

(Note: `check_access` is expecting a case - maybe we should refactor it to just take the domain)

**Review by snopoke — APPROVED** [review]

File: corehq/apps/userreports/views.py
Comment (stephherbers) [reply corehq/apps/userreports/views.py]: looks like there are two `check_access` and `access_check`: https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/registry/helper.py#L24 but I'm all for doing a refactor to use `access_check`

Comment (snopoke) [reply corehq/apps/userreports/views.py]: well that's confusing, I'll make a note to try and improve the interface

**Comment by stephherbers** [comment]
there is a test failure for corehq.apps.userreports.tests.test_pillow:AsyncIndicatorTest.[test_async_save_fails](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/userreports/tests/test_pillow.py#L778) that I'm still trying to figure out. After testing, appears that it unrelated to the `get_ucr_datasource_config_by_id` change

**Comment by stephherbers** [comment]
All tests passed and all QA tests passed they are just waiting on final approval
FYI @snopoke as reviewer

File: corehq/apps/userreports/reports/builder/forms.py
Comment (snopoke) [inline corehq/apps/userreports/reports/builder/forms.py:938]: I think this should be `commcare_project` shouldn't it? (Based on https://github.com/dimagi/commcare-hq/pull/30285/files#diff-eadcb42c8517e2f406bc980c0a5a354943556bbbb6602794faf2f33349fb5ae3R670)

File: corehq/apps/userreports/views.py
Comment (snopoke) [inline corehq/apps/userreports/views.py:747]: TODO

File: corehq/apps/userreports/views.py
Comment (snopoke) [inline corehq/apps/userreports/views.py:761]: TODO

**Comment by stephherbers** [comment]
@snopoke the TODO comments had the code changes in place to address them, so I removed those comments. I also agree with the change from domain to commcare_project https://github.com/dimagi/commcare-hq/pull/30285#:~:text=and%20not%20domain-,528350c,-Merge%20state

**Pull request merged**
